### DISCUSSION TITLE
antsMotionCorr split warp gives different output

### DISCUSSION THREAD

**Discussion by wschelle** [ANTsX/ANTs#1795] [Q&A] [answered]
Dear ANTsXperts,
I am using antsMotionCorr to create a motion warp displacement field:

antsMotionCorr  -d 3 \
-o [ $myfolder'output', $myfolder'output.nii.gz',$myfolder'output_avg.nii.gz'] \
-m MI[$myfolder'referencevolume.nii.gz', $myfolder'input.nii.gz', 1 , 32 , Regular, 0.25  ] \
-t Rigid[ 0.025 ] -u 1 -e 1 -s 0 -f 1 -i 20 -n 3 -w 1

This creates a 5D displacement field nifti ($myfolder'outputWarp.nii.gz') with shape (nX, nY, nZ, nT, 4). The last dimension (of shape 4) seems to represent the displacement warp field across X, Y, Z, and T, where the displacement along T (the 4th entry) equals to zero.
So far, so good.

Then, I want to split this displacement warp nifti up in nT separate displacement warp nifti's. Easily, this leads to me having nT nifti's of shape (nX, nY, nZ, 3).
The weird thing is, that I do not get the same result when I apply these separate warps compared to when I apply the warp in one go. Here's the code for the 2 different approaches:

Separate warps:

for (( i=0; i<nf; i++ ));
do
antsApplyTransforms -d 3 \
-r $myfolder'referencevolume.nii.gz' \
-t $myfolder\\${splitwarp[$i]} \
-i $myfolder\\${splitinput[$i]} \
-o \\$myfolder'splitoutput'$(printf "%04d" $i)'.nii.gz' \
-n LanczosWindowedSinc \
-v 1
done

In one go:

antsApplyTransforms -d 4 \
-r $myfolder'input.nii.gz' \
-t $myfolder'outputWarp.nii.gz' \
-i $myfolder'input.nii.gz' \
-o $myfolder'output_2.nii.gz' \
-n LanczosWindowedSinc \
-v 1

So, all this code runs smoothly (despite some backslashes not being visible here on the discuss page), but the output of the separate warp is different from the warp in one go. Different how? Well, after the separate warp, motion is corrected a little bit (i.e., less motion than the original input file), but more motion than the standard antsMotionCorr output or the 4d Warp applied in one go (these latter 2 lead to the same result).

I don't understand why the separate warp leads to a different result. Am I overlooking something? Help is greatly appreciated!

**Comment by cookpa** [comment]
What happens if you use `-n Linear`?

**Reply by wschelle** [reply]
Nothing changes to the problem when I set the interpolation type to Linear.

Also, I tried to warp the separate images manually using SimpleITK in Python, which resulted in exactly the same warped images as I would get from separately applying the warp via antsApplyTransforms. Note, that this also leads to too much motion. Thus, the 4D warp applied in one go does some magic that I don't -but would love to- understand.

Here's a link to the images I tested on:
https://drive.google.com/file/d/1d1rtfe-X72bY-_ZOymZMkQKLx39N8Qkg/view?usp=drive_link

**Reply by cookpa** [reply]
Thanks for the data. I was wondering if the Lanzcos was interpolating across time, which it can do with `antsApplyTransforms -d 4`. Linear could do the same, but shouldn't if the sample points are right in the middle of the time spacing. I will try when I can get a chance

**Reply by cookpa** [reply]
Are you using an older ANTs? The warp has no sform transform, which probably isn't the cause but is indicative of older ANTs that may be affected by other issues.

**Reply by wschelle** [reply]
Yes, I completely forgot to mention that I am running this on a server with ANTs Version: 2.2.0.dev137-g27f8e
There is no way I can upgrade quickly... But if that is the solution, then starting from which version this issue disappears (I believe they could not install the latest version)?

**Reply by cookpa** [reply]
OK good to know

**Comment by cookpa** [comment, marked as answer]
I think I found the cause: the split warps are 4D images, they should be 5D.

```
PrintHeader warpy/split/warp_split_0000.nii.gz 
 Spacing [0.903846, 0.903846, 0.9, 1]
 Origin [-94, 69.2638, -51.9875, 0]
 Direction 
1 0 0 0
0 -1 0 0
0 0 1 0
0 0 0 1

 Size : [208, 208, 102, 3]

  Image Dimensions   : [208, 208, 102, 3]
  Bounding Box       : {[-94 69.2638 -51.9875 0], [94 257.264 39.8125 3]}
  Voxel Spacing      : [0.903846, 0.903846, 0.9, 1]
  Intensity Range    : [-0.000134328, 0.000456632]
  Mean Intensity     : 0.000178321
  Direction Cos Mtx. : 
1 0 0 0
0 -1 0 0
0 0 1 0
0 0 0 1

  Voxel->RAS x-form  : 
  Image Metadata: 
    ITK_original_direction of unsupported type N3itk6MatrixIdLj4ELj4EEE
    ITK_original_spacing of unsupported type NSt3__16vectorIdNS_9allocatorIdEEEE
    ITK_sform_corrected = NO
    bitpix = 64
    cal_max = 0
    cal_min = 0
    datatype = 64
    dim[0] = 4
    dim[1] = 208
    dim[2] = 208
    dim[3] = 102
    dim[4] = 3
    dim[5] = 1
    dim[6] = 1
    dim[7] = 1
    dim_info = 0
    intent_code = 1007
    intent_p1 = 0
    intent_p2 = 0
    intent_p3 = 0
    nifti_type = 1
    pixdim[0] = 0
    pixdim[1] = 0.903846
    pixdim[2] = 0.903846
    pixdim[3] = 0.9
    pixdim[4] = 1
    pixdim[5] = 0
    pixdim[6] = 0
    pixdim[7] = 0
```

I think this is causing the warps to be misinterpreted, they should be 5D images with the vector components along the 5th dimension, like this:

```
    dim[0] = 5
    dim[1] = 208
    dim[2] = 208
    dim[3] = 102
    dim[4] = 1
    dim[5] = 3
    dim[6] = 1
    dim[7] = 1
    dim_info = 0
    intent_code = 1007
    intent_p1 = 0
    intent_p2 = 0
    intent_p3 = 0
    nifti_type = 1
    pixdim[0] = 0
    pixdim[1] = 0.903846
    pixdim[2] = 0.903846
    pixdim[3] = 0.9
    pixdim[4] = 1
    pixdim[5] = 1
    pixdim[6] = 0
    pixdim[7] = 0
```

Here's some code in antspy

```
import ants
import numpy as np

# Load the 5D NIfTI warp image
warp_image = ants.image_read('warpy/onego/warp.nii.gz')

# Convert the image to a numpy array
warp_array = warp_image.numpy()

# Loop through the 8 time points and check the fourth component (time) is zero
for i in range(warp_array.shape[3]):
    time_component = warp_array[:, :, :, i, 3]
    if not np.all(time_component == 0):
        print(f"Error: The fourth component (time) is not zero in time point {i}")
    else:
        print(f"Time point {i} passed time check (all zeros in the fourth component).")

for i in range(warp_array.shape[3]):
    # Extract the first 3 vector components (discard the 4th) for the current time point
    time_point_warp = warp_array[:, :, :, i, :3]

 
    # Create a multi-channel ANTsImage from the extracted 3D vectors
    time_point_warp_image = ants.from_numpy(time_point_warp, spacing=warp_image.spacing[0:3], origin=warp_image.origin[0:3], direction=warp_image.direction[:3,:3], has_components=True) 
    
    # Save the warp field transform
    ants.image_write(time_point_warp_image, f'output_warp_3D_{i}.nii.gz')
    
    print(f"3D warp field for time point {i} saved")
```

**Reply by wschelle** [reply]
Yes! That was it! Thanks for solving this so quickly!
So, to summarize (if people ever get stuck for similar reasons): the 4th dimension is reserved for time and the warps are in the 5th dimension (of a nifti file when using antsApplyTransorms, or SimpleITK).
